extensions/sql-database-projects/src/dialogs/addDatabaseReferenceDialog.ts

```diff
@@ -0,0 +1,403 @@
+/*---------------------------------------------------------------------------------------------
+ *  Copyright (c) Microsoft Corporation. All rights reserved.
+ *  Licensed under the Source EULA. See License.txt in the project root for license information.
+ *--------------------------------------------------------------------------------------------*/
+
+import * as azdata from 'azdata';
+import * as vscode from 'vscode';
+import * as constants from '../common/constants';
+
+import { Project, SystemDatabase, DatabaseReferenceLocation } from '../models/project';
+import { cssStyles } from '../common/uiConstants';
+import { IconPathHelper } from '../common/iconHelper';
+import { ISystemDatabaseReferenceSettings, IDacpacReferenceSettings } from '../models/IDatabaseReferenceSettings';
+import { isEmptyOrUndefined } from '../common/utils';
+
+export enum ReferenceType {
+	project,
+	systemDb,
+	dacpac
+}
+
+interface Deferred<T> {
+	resolve: (result: T | Promise<T>) => void;
+	reject: (reason: any) => void;
```

Comment:
Probably useful to have as a base abstract class things can extend from so it's easy to add to dialogs, as we work on adding UI to projects. (See https://github.com/microsoft/azuredatastudio/blob/main/extensions/arc/src/ui/components/initializingComponent.ts as an example)

Reply:
Ah yes that's a good idea. This PR is already pretty big, so I'll do this in a separate PR and also update the publish dialog to extend it.